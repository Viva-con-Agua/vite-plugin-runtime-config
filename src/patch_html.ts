import { ResolvedConfig } from "vite";
import { pickBy } from "lodash";

type ConfigLike = Pick<ResolvedConfig, "envPrefix" | "env">;

/**
 * Replace individual config keys in the given html string with values from the process environment.
 *
 * Individual keys are written using the syntax `{{ VITE_KEY }}`.
 */
export function replaceIndividualKeys(html: string, vite_cfg: ConfigLike): string {
    const regexp = new RegExp(`{{ *(${vite_cfg.envPrefix || "VITE_"}[\\w_]+?) *}}`, "ig");
    return html.replaceAll(regexp, (match, ...args: any[]) => {
        const key = args[0];
        if (vite_cfg.env[key] == undefined) {
            console.warn(
                `[WARNING] index.html references runtime config key ${key} which is undefined, using empty string`
            );
        }
        return vite_cfg.env[key] || "";
    });
}

/**
 * Replace references to the complete runtime config in the given html string with a rendered javascript object literal of the config.
 *
 * Those references are written using the syntax `{% VITE_RT_CONFIG %}`
 */
export function replaceCompleteConfig(html: string, vite_cfg: ConfigLike): string {
    const regexp = new RegExp(`{% *${vite_cfg.envPrefix || "VITE_"}RT_CONFIG *%}`, "ig");
    return html.replaceAll(regexp, renderCompleteConfig(vite_cfg));
}

export function renderCompleteConfig(vite_cfg: ConfigLike): string {
    return JSON.stringify(
        pickBy(vite_cfg.env, (_, key) => key.startsWith((vite_cfg.envPrefix as string) || "VITE_"))
    );
}
