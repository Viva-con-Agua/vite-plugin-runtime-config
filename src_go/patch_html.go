package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"regexp"
	"strings"
)

// ReplaceIndividualKeys in the given html string with values from the process environment.
func ReplaceIndividualKeys(html string, prefix *string) string {
	regex := fmt.Sprintf("{{ *%s[\\w_]+? *}}", *prefix)
	return regexp.MustCompile(regex).ReplaceAllStringFunc(html, func(fullMatch string) string {
		cfgKey := regexp.MustCompile("[\\w_]+").FindString(fullMatch)
		value, found := os.LookupEnv(cfgKey)
		if !found {
			log.Printf("Configuration value %s is undefined. Using empty string\n", cfgKey)
		}

		return value
	})
}

// ReplaceCompleteConfig replaces references to the complete runtime config in the given html string with a rendered
// javascript object literal of the config.
func ReplaceCompleteConfig(html string, prefix *string) string {
	regex := fmt.Sprintf("{%% *%sRT_CONFIG *%%}", *prefix)
	return regexp.MustCompile(regex).ReplaceAllString(html, marshalEnvironment(prefix))
}

// marshal the process's environment variables into JSON so that it can be put into html
func marshalEnvironment(prefix *string) string {
	cfgMap := make(map[string]string)
	for _, envString := range os.Environ() {
		if strings.HasPrefix(envString, *prefix) {
			parts := strings.SplitN(envString, "=", 2)
			cfgMap[parts[0]] = parts[1]
		}
	}
	encodedCfgMap, _ := json.Marshal(cfgMap)
	return string(encodedCfgMap)
}
