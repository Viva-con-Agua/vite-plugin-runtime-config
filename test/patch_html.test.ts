import { test, expect, beforeAll } from "vitest";
import { readFile } from "node:fs/promises";
import * as Path from "path";
import * as patch_html from "@/patch_html";

let html: string;
const mock_cfg: patch_html.ConfigLike = {
    envPrefix: "VITE_",
    env: {
        VITE_TEST_IN_HEAD: "test-in-head-value",
        VITE_TEST_PERSON_NAME: "max mustermann",
    },
};

beforeAll(async () => {
    html = await readFile(Path.join(__dirname, "test.html"), {
        encoding: "utf8",
    });
});

test("replaceIndividualKeys", () => {
    expect(patch_html.replaceIndividualKeys(html, mock_cfg)).toMatchSnapshot();
});

test("replaceCompleteConfig", () => {
    expect(patch_html.replaceCompleteConfig(html, mock_cfg)).toMatchSnapshot();
});
