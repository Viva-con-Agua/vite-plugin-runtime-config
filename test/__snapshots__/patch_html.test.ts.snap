// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replaceCompleteConfig 1`] = `
"<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="test" content="{{ VITE_TEST_IN_HEAD }}" />
        <title>Vite Test html</title>
    </head>
    <body>
        <h1>Hello {{ VITE_TEST_PERSON_NAME }}</h1>
        <script type="application/ecmascript">
            window.config = Object.freeze({"VITE_TEST_IN_HEAD":"test-in-head-value","VITE_TEST_PERSON_NAME":"max mustermann"});
        </script>
    </body>
</html>
"
`;

exports[`replaceIndividualKeys 1`] = `
"<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta name="test" content="test-in-head-value" />
        <title>Vite Test html</title>
    </head>
    <body>
        <h1>Hello max mustermann</h1>
        <script type="application/ecmascript">
            window.config = Object.freeze({% VITE_RT_CONFIG %});
        </script>
    </body>
</html>
"
`;
